### PR TITLE
osd: Implement asynchronous recovery sleep

### PR DESCRIPTION
**Pull request by neha-ojha**
Rather than blocking the main op queue, do an asynchronous recovery sleep.

Signed-off-by: Neha Ojha <nojha@redhat.com>

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -8881,18 +8889,35 @@ void OSD::do_recovery(
...
-    handle.reset_tp_timeout();
+  if (cct->_conf->osd_recovery_sleep > 0 && service.recovery_needs_sleep) {
+    auto recovery_requeue_callback = new FunctionContext([this, pg](int r) {
+      pg->lock();
+      dout(20) << "do_recovery wake up at "
+               << ceph_clock_now()
+	       << ", re-queuing recovery" << dendl;
+      service.recovery_needs_sleep = false;
+      pg->recovery_queued = false;
+      pg->queue_recovery();
```
Comment (jdurgin) [inline src/osd/OSD.cc]: At this point this PGQueuable already went through and was accounted for in recovery_ops_reserved when it was originally enqueued, but pg->queue_recovery() will end up calling OSDService::queue_for_recovery()
->OSDService::_maybe_queue_recovery() which will try to account it in recovery_ops_reserved again.

It seems like we can avoid taking the pg lock and changing any pg state here, and re-add the pg directly via OSDService::_queue_for_recovery() under the recovery_lock.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -8881,18 +8889,35 @@ void OSD::do_recovery(
...
-  if (cct->_conf->osd_recovery_sleep > 0) {
-    handle.suspend_tp_timeout();
-    pg->unlock();
-    utime_t t;
-    t.set_from_double(cct->_conf->osd_recovery_sleep);
-    t.sleep();
-    dout(20) << __func__ << " slept for " << t << dendl;
-    pg->lock();
-    handle.reset_tp_timeout();
+  if (cct->_conf->osd_recovery_sleep > 0 && service.recovery_needs_sleep) {
```
Comment (jdurgin) [inline src/osd/OSD.cc]: I think your requeuing algorithm makes sense - could you add a comment here explaining it for future code readers?

**Comment by jdurgin** [comment]
right now nothing in the docs mentions this option - it should be added to doc/rados/configuration/osd-config-ref.rst

I'm thinking the default for this setting should be a small value to (a) get tested with all the existing suites and (b) impact client i/o less during default recovery. Something like 0.01 perhaps? This is one of those settings where different defaults for ssds vs hdds would be ideal.

File: src/osd/OSD.cc (hunk 1 above)
Comment (neha-ojha) [inline src/osd/OSD.cc]: I'll try making these changes and let you know how it goes.

File: src/osd/OSD.cc (hunk 2 above)
Comment (neha-ojha) [inline src/osd/OSD.cc]: sure.

**Pull request merged**